src/operator/numpy/np_matmul_op-inl.h

```diff
@@ -0,0 +1,356 @@
+
+template<int ndim>
+mshadow::Shape<ndim> GetStride(mshadow::Shape<ndim> shape, size_t N) {
+  /*!
+   * \brief Calculate stride of each dim from shape 
+   */
+  mshadow::Shape<ndim>stride;
+  size_t tmp = 1;
+  for (int i = N - 1; i >= 0; --i) {
+    stride[i] = tmp;
+    tmp *= shape[i];
+  }
+  return stride;
+}
+
+template<int ndim>
+mshadow::Shape<ndim> GetKernelShape(const mxnet::TShape& shape,
+                                    size_t N, bool T = false) {
+  /*!
+   * \brief Get mshadow::Shape from mxnet::TShape. Extra dims is filled with 1.
+   * \param N - ndim of mshape::Shape shape.
+   * \param T - If T is True, transpose the last two axis, otherwise not.
+   */
+  mshadow::Shape<ndim>k_shape;
+  for (int i = shape.ndim() - 1, j = N - 1; i >= 0 || j >= 0 ; --i, --j) {
+    if (i >= 0) {
+      k_shape[j] = shape[i];
+    } else {
+      k_shape[j] = 1;
+    }
+  }
+  if (T) {  // transpose the latest two axes
+    size_t t = k_shape[N - 1];
+    k_shape[N - 1] = k_shape[N - 2];
+    k_shape[N - 2] = t;
+  }
+  return k_shape;
+}
+
+template<int ndim>
+mshadow::Shape<ndim> BroadcastKernelShape(mshadow::Shape<ndim> in_shape,
+                                          mshadow::Shape<ndim> broadcast_shape,
+                                          size_t N, size_t* size) {
+  /*!
+   * \brief Broadcast in_shape(ndim = N) to broadcast_shape(ndim = N) expect the last two axes.
+            Make sure that: If i < N - 2 and in_shape[i] != broadcast_shape[i], in_shape[i] == 1.
+   * \param N - ndim of both in_shape and broadcast_shape.
+   * \param size - The size of the broadcast_shape.
+   */
+  mshadow::Shape<ndim>out_shape(in_shape);
+  *size = 1;
+  for (size_t i = 0; i < N - 2; ++i) {
+    out_shape[i] = std::max(in_shape[i], broadcast_shape[i]);
+    *size *= out_shape[i];
+  }
+  *size *= (out_shape[N - 2] * out_shape[N - 1]);
+  return out_shape;
+}
+
+template<int req>
+struct NDMatmul {
+  /*!
+   * \brief matmul(a, b) in both N-D(N >= 2) case.
+            It is treated as a stack of matrices residing in the last two indexes and broadcast accordingly.
+   * \param out - output: insert 'value' to 'arr' according to 'index'.
+   * \param a - input: the first argument.
+   * \param b - input: the second argument.
+   * \param ndim - ndim of a, b and output. Because of broadcast, regard their ndim as equal.  
+   */
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* out,
+                                  const DType* a, const DType* b,
+                                  const mshadow::Shape<10> a_stride,
+                                  const mshadow::Shape<10> b_stride,
+                                  const mshadow::Shape<10> out_stride,
+                                  const mshadow::Shape<10> a_shape,
+                                  const mshadow::Shape<10> b_shape,
+                                  const mshadow::Shape<10> out_shape,
+                                  const size_t ndim){
```

Comment:
You may refer to the implementation of batch_dot. There is no need to store the strides and shapes as Shape<10>. You can reshape the array to a 3D array and just dot the last two dimensions:
https://github.com/apache/incubator-mxnet/blob/1aa1b5a9ab53bb57a3c653793fb824d01f2d5e81/src/operator/tensor/dot-inl.h#L1348-L1409

Reply:
Thanks for your advice. But, may I ask how could I broadcast the shape if I don't store strides and shapes? 
e.g. Matrix A in shape (2, 1, 3, 4, 5) and matrix B in shape (3, 1, 5, 2), C=np.matmul(A, B) would broadcast A and B to shape (2, 3, 3, 4, 5) and (3, 3, 5, 2) respectively, and C's shape would be (2, 3, 3, 4, 2).
If I didn't  store shape and stride, I think I should copy the content in each array to get the consistent shape firstly, and use your method then.  Is this your mean?

Reply:
> You may refer to the implementation of batch_dot. There is no need to store the strides and shapes as Shape<10>. You can reshape the array to a 3D array and just dot the last two dimensions:
> 
> https://github.com/apache/incubator-mxnet/blob/1aa1b5a9ab53bb57a3c653793fb824d01f2d5e81/src/operator/tensor/dot-inl.h#L1348-L1409

Thanks for your advice. But, may I ask how could I broadcast the shape if I don't store strides and shapes? 
e.g. Matrix A in shape (2, 1, 3, 4, 5) and matrix B in shape (3, 1, 5, 2), C=np.matmul(A, B) would broadcast A and B to shape (2, 3, 3, 4, 5) and (3, 3, 5, 2) respectively, and C's shape would be (2, 3, 3, 4, 2).
If I didn't  store shape and stride, I think I should copy the content in each array to get the consistent shape firstly, and use your method then.  Is this your mean?